lib/rucio/core/did.py

```diff
@@ -3002,3 +3002,53 @@ def insert_deleted_dids(filter_: "ColumnExpressionArgument[bool]", *, session: "
             is_archive=did.is_archive,
             constituent=did.constituent
         ).save(session=session, flush=False)
+
+@transactional_session
+def add_files_with_structure(
+        files: "Iterable[dict[str, Any]]",
+        account: "InternalAccount",
+        *,
+        session: "Session"
+):
+    """
+    Add files with structure
+
+    :param files: An iterable of files to add. It should be a list of dictionaries with the following keys:
+        - scope: The scope name.
+        - name: The data identifier name.
+        - rse: The RSE name.
+        - pfn: The physical file name.
+        - bytes: The file size in bytes.
+        - adler32: The adler32 checksum.
+    :param account: The account to use.
+    :param session: The database session in use.
+    """
+
+    for file in files:
+        if "scope" not in file:
+            raise exception.InputValidationError("Missing 'scope' in file")
+        if "name" not in file:
+            raise exception.InputValidationError("Missing 'name' in file")
+        scope = file["scope"]
+        name = file["name"]
+        parts = name.split("/")
+        directories = [
+            "/".join(parts[:i + 1]) for i in range(len(parts) - 1)
+        ]
```

Comment:
It would be good to have a comment to explain and help visualize what this does, e.g.

```suggestion
        # get parent paths (e.g. 'a/b/c/d' -> ['a', 'a/b', 'a/b/c'])
        parts = name.split("/")
        directories = [
            "/".join(parts[:i + 1]) for i in range(len(parts) - 1)
        ]
```